src/ui/VirtualFileRenderer/VirtualFileRenderer.tsx

```diff
@@ -0,0 +1,323 @@
+import { useWindowVirtualizer } from '@tanstack/react-virtual'
+// eslint-disable-next-line no-restricted-imports
+import type { Dictionary } from 'lodash'
+import Highlight, { defaultProps } from 'prism-react-renderer'
+import { useEffect, useLayoutEffect, useRef } from 'react'
+import { useHistory, useLocation } from 'react-router-dom'
+
+import { requestAnimationTimeout } from 'shared/utils/animationFrameUtils'
+import { cn } from 'shared/utils/cn'
+import { prismLanguageMapper } from 'shared/utils/prismLanguageMapper'
+
+import { ColorBar } from './ColorBar'
+
+// copied from prism-react-renderer since they don't export it
+type Token = {
+  types: string[]
+  content: string
+  empty?: boolean
+}
+
+interface CodeBodyProps {
+  tokens: Token[][]
+  getLineProps: Highlight['getLineProps']
+  getTokenProps: Highlight['getTokenProps']
+  coverage: Dictionary<'H' | 'M' | 'P'>
+  codeDisplayOverlayRef: React.RefObject<HTMLDivElement>
+}
+
+const CodeBody = ({
+  tokens,
+  getLineProps,
+  getTokenProps,
+  coverage,
+  codeDisplayOverlayRef,
+}: CodeBodyProps) => {
+  const history = useHistory()
+  const location = useLocation()
+
+  const initialRender = useRef(true)
+  const wrapperRef = useRef<HTMLDivElement>(null)
+
+  // virtualize code lines
+  const virtualizer = useWindowVirtualizer({
+    count: tokens.length,
+    estimateSize: () => 18,
```

Comment:
I would add a comment or pull this into a const
```suggestion
    estimateSize: () => 18, // line row height
```